File: bridge/bridge.java
// Abstraction
interface Renderer {
    void renderCircle(float radius);
}

// Concrete Implementations
class OpenGLRenderer implements Renderer {
    @Override
    public void renderCircle(float radius) {
        System.out.println("Drawing a circle with OpenGL with radius " + radius);
    }
}

class DirectXRenderer implements Renderer {
    @Override
    public void renderCircle(float radius) {
        System.out.println("Drawing a circle with DirectX with radius " + radius);
    }
}

// Refined Abstraction
class Circle {
    private float radius;
    private Renderer renderer;

    public Circle(float radius, Renderer renderer) {
        this.radius = radius;
        this.renderer = renderer;
    }

    public void draw() {
        renderer.renderCircle(radius);
    }
}

// Client code
public class Main {
    public static void main(String[] args) {
        Renderer openGL = new OpenGLRenderer();
        Renderer directX = new DirectXRenderer();

        Circle circle1 = new Circle(5, openGL);
        Circle circle2 = new Circle(10, directX);

        circle1.draw();
        circle2.draw();
    }
}

File: bridge/bridge.py
# Avoid creating OpenGLCircle, DirectXCircle, etc
from typing import Protocol

# Define Renderer protocol (Abstraction)
class Renderer(Protocol):
    def render_circle(self, radius: float) -> None:
        pass

# ConcreteImplementor 1
class OpenGLRenderer:
    def render_circle(self, radius: float) -> None:
        print(f"OpenGL: Rendering circle with radius {radius}")

# ConcreteImplementor 2
class DirectXRenderer:
    def render_circle(self, radius: float) -> None:
        print(f"DirectX: Rendering circle with radius {radius}")

# Refined Abstraction
class Circle:
    def __init__(self, renderer: Renderer, radius: float) -> None:
        self.renderer = renderer
        self.radius = radius

    def draw(self) -> None:
        self.renderer.render_circle(self.radius)

# Client code
opengl = OpenGLRenderer()
directx = DirectXRenderer()

circle1 = Circle(opengl, 5)
circle2 = Circle(directx, 10)

circle1.draw()
circle2.draw()
